agent/engine/dockerstate/docker_task_engine_state.go

```diff
@@ -143,6 +146,22 @@ func (state *DockerTaskEngineState) ContainerMapByArn(arn string) (map[string]*a
 	return ret, ok
 }
 
+// TaskByShortID retrieves the task of a given docker short container id
+func (state *DockerTaskEngineState) TaskByShortID(cid string) ([]*api.Task, bool) {
+	state.lock.RLock()
+	defer state.lock.RUnlock()
+
+	var tasks []*api.Task
+	for id := range state.idToTask {
+		if strings.HasPrefix(id, cid) {
+			task, _ := state.TaskByID(id)
```

Comment:
@richardpen please correct me if I'm wrong here. But, IIUC, this has the potential to cause a deadlock as an intervening write lock between line 151 (`state.lock.RLock()`) and this line has the potential to cause a deadlock. 

Please see [RWMutex](https://golang.org/pkg/sync/#RWMutex) for more details.

I think something like this is a better alternative:
```
func (state *DockerTaskEngineState) GetAllContainerIDs() []string {
    state.lock.RLock()
    defer state.lock.RUnlock()

    var ids []string
    for id := range state.idToTask {
        ids = append(ids, id)
    }

    return ids
}

func (state *DockerTaskEngineState) TaskByShortID(cid string) ([]*api.Task, bool) {
    containerIDs := state.GetAllContainerIDs()
    var tasks []*api.Task
    for id := range containerIDs {
        if strings.HasPrefix(id, cid) {
            if task, ok := state.TaskByID(id); ok {
                tasks = append(tasks, task)
            }
        }
    }

    return tasks, len(tasks) > 0
}